fix!: Sync deleted doctypes on migration

niraj2477 opened the pull request:
Fixes part of : https://github.com/frappe/frappe/issues/18114

niraj2477 committed:
fix: Delete stale doctypes, reports and pages

niraj2477 committed:
refactor: Fix ambigious var name

niraj2477 committed:
fix: Update get_doc filter

niraj2477 commented:
@ankush Can you please help me how should I write test cases for this fix? I have tested the fix via `post schema sync`  patch, but not sure if it is the right way to test this.

ankush commented on frappe/model/sync.py:
```diff
@@ -135,3 +136,89 @@ def get_doc_files(files, start_path):
+def remove_stale_reports():
+	"""Find and remove any stale reports."""
+
+	if not frappe.get_site_config().get("developer_mode"):
+		# Skip if developer mode is not enabled.
+		print("Enable developer mode to check for stale reports.")
```
Is it risky to run this in prod? it only touches non-standard reports.

niraj2477 replied:
- If we look at the `on_trash` method for report DOCTYPE, we are not allowed to delete reports if we are not in `developer mode`  or running a `patch`.
https://github.com/frappe/frappe/blob/develop/frappe/core/doctype/report/report.py#L85
- If we try to delete reports in production, validation exception will be raised and migration will fail.

ankush replied:
Umm, we can modify that logic.

niraj2477 commented on frappe/model/sync.py:
```diff
@@ -135,3 +136,89 @@ def get_doc_files(files, start_path):
+		print("Enable developer mode to check for stale reports.")
+		return
+
+	reports_names = frappe.get_all(
```
We will be checking for standard reports, if not in production.

niraj2477 commented on frappe/model/sync.py:
```diff
@@ -135,3 +136,89 @@ def get_doc_files(files, start_path):
+
+	reports_names = frappe.get_all(
+		"Report", filters={"is_standard": "Yes"}, fields=["name", "module"]
+	)
```
We will be checking for standard reports, if not in production.

ankush committed:
ci: update patch workflow

- start with v13 DB, drop py3.7
- New flow v13DB -> V14 -> V15 (when released) -> PR commit
- Restart bench before final migrate

ankush commented:
@niraj2477 I'll write some custom test for this in migration test workflow. 


Reference implementation: https://github.com/frappe/erpnext/pull/27703

ankush committed:
refactor: stale object removal

- Do it after migration, post migration might use some old doctype.
- stale -> orpahn
- dont commit changes individually
- Remove table doctype too

ankush commented on frappe/model/sync.py:
```diff
@@ -135,3 +136,92 @@ def get_doc_files(files, start_path):
+	So this is supposed to be non-destrictive operation.
+	"""
+
+	doctype_names = frappe.get_all("DocType", {"istable": 0, "custom": 0}, pluck="name")
```
```suggestion
	doctype_names = frappe.get_all("DocType", {"istable": 0, "custom": 0}, pluck="name")
```

Any reason why child tables are ignored?

niraj2477 replied:
- If we delete child tables, there is possibility where these tables are added as `table` field in some other DOCTYPE.

ankush replied:
Without controller files these are useless anyway. They will give `ImportError`.

niraj2477 replied:
True. We can still delete them, doing so will throw an error that `doctype has been deleted.`

ankush committed:
test: orpahned doctype check after patch test

ankush commented:
TODO:

- [x]  Run ERPNext tests with this branch. https://github.com/frappe/erpnext/pull/37104
- [x] This fails when module is deleted.

ankush commented:
Tried to write a test but modules get cached by python interpreter even if they are gone from disk. Kinda hard to do this :grimacing:

ankush committed:
fix: Ignore all non-import exceptions

niraj2477 commented:
Can we try removing module entries from the `sys.modules`? 🤔

ankush commented:
> Can we try removing module entries from the sys.modules? 🤔 

Or spawn a new process for it :grimacing:

ankush commented:
IMO lets start with doctype first, if things go fine I'll merge the rest later.

ankush committed:
revert: syncing of report and pages
